### ISSUE TITLE
Option to drag by fixed amounts/intervals

### ISSUE THREAD

**Issue by adrianbj** [nathancahill/Split.js#116]
Thanks for a great script. 
Just wondering if you would consider an option to snap dragging to a defined interval of pixels.

I am using this in conjunction with Ace editor and it would be great to be able to set the `dragInterval` to the same height as the Ace editor `lineHeight` setting. 

I am sure there could be other use cases as well for this functionality.

Thanks!

**Comment by nathancahill** [comment]
Yes, I'd accept a PR for this idea.

**Comment by adrianbj** [comment]
OK, sounds good. I have already implemented here, although I did it outside the splitjs codebase using `onDragEnd()` so I'll need to rework it bit, but I'll take a look.

**Comment by nathancahill** [comment]
Look at the code that uses `getOption` to pass the option in, then the actual drag code is in the `drag` function here: https://github.com/nathancahill/Split.js/blob/master/src/split.js#L261

**Comment by adrianbj** [comment]
Actually, I am using  `+ this[bGutterSize]` but should I use ` + this[aGutterSize]` or does there need to be more logic here to know when to use which? In my testing they are always the same and half of the specified `gutterSize`.

**Comment by nathancahill** [comment]
Great! This looks good. You don't need to account for the gutter in the dragging math, that's handled later. I would say subtract half of the dragInterval, so that the gutter is more centered under the cursor. Change that line to this:

```js
(Math.ceil(offset / snapInterval) * snapInterval) - (snapInterval / 2)
```

You can also get rid of this if statement and the `snappedInterval` variable:

```js
if(offset!== snappedInterval) {
```
And just set offset directly instead:

```js
offset = (Math.ceil(offset / snapInterval) * snapInterval) - (snapInterval / 2)
```

Please add the new option to the README, both in the markdown table and further down with the default value and description.

I'll handle building and minifying, so the only changes in the PR should be in `src/split.js`. Thanks!

**Comment by nathancahill** [comment]
Also I think dragInterval is a better name since we already use `snap` for snapping to minSize.

**Comment by nathancahill** [comment]
One more thing for your PR, add yourself to the AUTHORS.md file.

**Comment by adrianbj** [comment]
Thanks for the feedback.

> I would say subtract half of the dragInterval, so that the gutter is more centered under the cursor. Change that line to this:
(Math.ceil(offset / snapInterval) * snapInterval) - (snapInterval / 2)

I don't think this works. It's actually half the gutter width that works for me - at least in my setup. Take a look at these screencasts.

This code:
```
        if (dragInterval !== 0) {
            var draggedInterval = (Math.ceil(offset / dragInterval) * dragInterval) + this[aGutterSize];
            if(offset!== dragInterval) {
                offset = draggedInterval;
            }
        }
```
results in:
![g1m71doaoq](https://user-images.githubusercontent.com/871211/46705651-2eeb8c80-cc7c-11e8-84c1-fabe03e3aac0.gif)

But if I convert to what I think you are suggesting:
```
        if(snapInterval !== 0) {        
            offset = (Math.ceil(offset / dragInterval) * dragInterval) - (dragInterval / 2);
        }
```
I end up with this where the divider is not correctly positioned between the lines:
![mdi8ptv1m3](https://user-images.githubusercontent.com/871211/46705892-555df780-cc7d-11e8-9233-63b58a328a02.gif)

A simplified working version for me is:
```
        if(dragInterval !== 0) {
            offset = (Math.ceil(offset / dragInterval) * dragInterval) + this[aGutterSize];
        }
```

By the way, I am setting the new `dragInterval` value to match the ACEEditor lineHeight. I am also setting splitjs's `minSize` to this same value.

What do you think of the revised version still using the `+ this[aGutterSize]` ? Or is there something I am missing in getting the ` - (dragInterval / 2)` version to work as expected?

Thanks!

**Comment by nathancahill** [comment]
I see, your version matches the line heights. I was optimizing for keeping the cursor over the gutter. You can see the difference in the two (very nice) gifs you posted.

**Comment by adrianbj** [comment]
> I see, your version matches the line heights. I was optimizing for keeping the cursor over the gutter. You can see the difference in the two (very nice) gifs you posted.

Right - I do now see why you wanted the mouse cursor over the gutter, but I think that the position of the gutter relative to the line height is more important - otherwise the end result is quite ineffective. 

This seems to accommodate both needs relatively well:
```
        if(dragInterval !== 0) {
            offset = (Math.ceil(offset / dragInterval) * dragInterval) - dragInterval + this[aGutterSize];
        }
```
It still jumps obviously, but at least now each time it snaps it snaps into place under the cursor.

What do you think - would you be happy accepting this version?

**Comment by nathancahill** [comment]
Oh, here's the better option:

```
offset = Math.round(offset / dragInterval) * dragInterval
```

`round` instead of `ceil`. Keeps it centered and at the correct offsets for your rows.

**Comment by nathancahill** [comment]
Sorry, I'm not trying to be difficult, just want to make sure this new option works all around and not just for this specific use case. In your example, the "correct" sizing will only apply to the first (top) element of the split. If you tried your example in the second (bottom) element, the sizing would be off because `aGutterSize` is for the top element.

I'm trying to come up with a solution that doesn't use `aGutterSize` or `bGutterSize` because those are specific to 1 of the 2 elements.

**Comment by nathancahill** [comment]
I thought about this a bit more, and I think we'll need a second option besides `dragInterval`. I'm thinking `gutterAlign`, which can have the values `'start'`, `'end'` or `'center'`. This determines if the gutter sits before, after or in between the elements. The current behavior is `'center'` and what you're needing for Ace is `'start'` or `'end'`.

Here's the 3 behaviors against a 20px background, with `gutterSize: 20` and `dragInterval: 20`.

![](http://g.recordit.co/6UIeAHYMMR.gif)
![](http://g.recordit.co/a6WS7chx7N.gif)
![](http://g.recordit.co/cwSA8s8D65.gif)

**Comment by nathancahill** [comment]
In `'start'` and `'end'` you can see how the gutter leads or trails the mouse. In `'center'` it's directly under it.

**Comment by nathancahill** [comment]
Added this changes in PR #155. Please give it a try and see if that works for your use case.

**Comment by adrianbj** [comment]
That looks good and seems like a great solution - thanks for picking up on this and making it some much better / more versatile. 

I tested the PR you just posted and the 'end' option works beautifully!

**Comment by nathancahill** [comment]
Merged and published as 1.5.3. Thanks!

**Comment by adrianbj** [comment]
Thank you very much!

**Comment by adrianbj** [comment]
Just a thought, but I wonder if it might be more semantically correct for the default value to be `1` instead of `0`. An actual value of `0` would mean that dragging wouldn't move the split at all. `1` would be the equivalent of dragging naturally / smoothly one pixel at a time.

I don't feel strongly about it one way or the other, but thought I should bring it up given that my first code example used `0` so I put us on this path to start with :)

**Comment by nathancahill** [comment]
Yeah, good observation. Want to submit that as a PR? I think we just have to change the default and the `if (dragInterval > 0) {` check.

**Comment by adrianbj** [comment]
Sure, happy to. Just one question - what should happen if `dragInterval` is actually set to `0`. Do you think it makes sense to have this as a way to prevent a gutter from being draggable at all?

Maybe this could actually be a useful option in some scenarios, especially when there are more than two elements, or as a way to lock dragging after some user interaction, etc.

What do you think?

**Comment by nathancahill** [comment]
I'd probably ignore `0`. Seems a little unexpected to have different behavior like that, even if it is mathematically correct.

**Comment by nathancahill** [comment]
Published in v1.5.4.
